apps/dashboard/src/utils/liquid-autocomplete.ts

```diff
@@ -1,159 +1,114 @@
 import { LiquidVariable } from '@/utils/parseStepVariablesToLiquidVariables';
 import { CompletionContext, CompletionResult } from '@codemirror/autocomplete';
+import { TRANSFORMERS } from '@/components/primitives/field-editor/variable-popover/constants';
+
+interface CompletionOption {
+  label: string;
+  type: string;
+  boost?: number;
+}
+
+function isInsideLiquidBlock(beforeCursor: string): boolean {
+  const lastOpenBrace = beforeCursor.lastIndexOf('{{');
+  const lastCloseBrace = beforeCursor.lastIndexOf('}}');
+
+  return lastOpenBrace !== -1 && lastOpenBrace > lastCloseBrace;
+}
+
+function getContentAfterPipe(content: string): string | null {
+  const pipeIndex = content.lastIndexOf('|');
+  if (pipeIndex === -1) return null;
+
+  return content.slice(pipeIndex + 1).trimStart();
+}
+
+function getFilterCompletions(afterPipe: string): CompletionOption[] {
+  return TRANSFORMERS.filter((f) => f.label.toLowerCase().startsWith(afterPipe.toLowerCase())).map((f) => ({
+    label: f.value,
+    type: 'function',
+  }));
+}
+
+function getVariableCompletions(
+  word: { text: string; from: number; to?: number } | null,
+  pos: number,
+  variables: LiquidVariable[]
+): CompletionResult | null {
+  if (!word && !variables.length) return null;
+
+  if (word?.text.startsWith('payload.')) {
+    const payloadPath = word.text.slice(8);
+    const existingVariables = variables.filter((v) => v.label.startsWith('payload.'));
```

Comment:
Is this logic here to support autocompleting arbitrary variables that are not synced with the API yet? So far, the logic is that existing variables are sent from the server inside each step.

So, we need to use the ones from the server and combine them with the "draft" variables on the client side.